### PR TITLE
[RDY] Fix stats editing menu bug

### LINKED ISSUE #17889: Debug menu: Various bugs.
Found quite a bit of unexpected behaviours using the debug menu.
- [x] When you use the debug menu to edit a player/npc's stats and use escape to cancel the number input screen, it sets the stat to zero.
- [x] When escaping from the weather selection menu you get NULL weather.

(I didn't test all escape key menu's btw. I'm only looking in all of them to add them to the wiki).
- [x] The 'test item group' menu can only be exited by selecting the last 'cancel' field. No escape key or 'q' key.
- [ ] Both 'Show sound clustering' and 'display weather' don't seem to do anything useful.

### PR DESCRIPTION
**Pull request by ghost**
This fixes the first problem pointed out in issue 17889 , also fixes two conflicting menu entry letters.

> When you use the debug menu to edit a player/npc's stats and use escape to cancel the number input screen, it sets the stat to zero.

### CODE REVIEW

**Comment by mugling** [comment]
Does this fix all of #17889?

**Comment by DLaboratory** [comment]
oops, the author edited it but I was unaware. This only fixes the first bug Soyweiser listed.

**Comment by mugling** [comment]
@DLaboratory please update your preamble to avoid automatic closure of the issue.

@Soyweiser please one problem per issue and restrict edits to clarification etc and mark them with `EDIT` or something to that effect.

File: src/output.cpp [hunk 1]
```diff
@@ -713,6 +713,16 @@ int query_int( const char *mes, ... )
     return atoi( raw_input.c_str() );
 }
 
+std::string query_string( const char *mes, ... )
```
Comment (mugling) [inline src/output.cpp]: Why not the existing `string_format`?

Comment (DLaboratory) [inline src/output.cpp]: I thought to create a new function `query_string` performs quite similar to existing `query_int`.

Comment (mugling) [inline src/output.cpp]: Could be trivially replaced by `string_input_popup( string_format( ...`

Comment (mugling) [inline src/output.cpp]: Alternatively refactor `bool query_int( std::string msg, bool &result )` where the return is whether the value was set. There aren't many uses to update and almost all are within the debug menu code.

Comment (DLaboratory) [inline src/output.cpp]: OK, I will refactor `query_int` later.

**Comment by DLaboratory** [comment]
Ready.

File: src/game.cpp [hunk 2]
```diff
@@ -4166,12 +4166,9 @@ void game::debug()
...
 
                         if( bp_ptr != nullptr ) {
-                            std::string raw_input = query_string( "Set the stat to? Currently: %d", *bp_ptr );
-                            if( raw_input.empty() ) { // ESC is pressed
-                                break;
-                            }
-                            int value = atoi( raw_input.c_str() );
-                            if( value >= 0 ) {
+                            int value;
+                            bool got_value = query_int( value, "Set the stat to? Currently: %d", *bp_ptr );
```
Comment (mugling) [inline src/game.cpp]: You don't need the temporary. You can do `if( query_int(`. This is the same idiom used by `JsonObject::read`

File: src/output.cpp (hunk 1 above)
Comment (BevapDin) [inline src/output.cpp]: > bool query_int( std::string msg, bool &result ) 

A return value of `std::pair<int, bool>` looks better. It avoids the temporary bool parameter.

```C++
bool b;
const int v = query_int(..., &b);
if(v) {
    return v;
}
```
vs
```C++
const int p = query_int(...);
if(p.second) {
    return p.first;
}
```

This style is also used by the standard library, for example in `std::map<T>::insert`

File: src/game.cpp (hunk 2 above)
Comment (DLaboratory) [inline src/game.cpp]: Remembered wrongly. I used to think `if( function() )` style isn't accepted in this project.

File: src/output.cpp [hunk 3]
```diff
@@ -700,17 +700,21 @@ bool internal_query_yn( const char *mes, va_list ap )
...
     const std::string text = vstring_format( mes, ap );
     va_end( ap );
 
     std::string raw_input = string_input_popup( text );
+    if( raw_input.empty() ) { //ESC key is pressed
+        return false;
+    }
 
     //Note that atoi returns 0 for anything it doesn't like.
-    return atoi( raw_input.c_str() );
```
Comment (mugling) [inline src/output.cpp]: You should catch an invalid entry here and return `false`

File: src/iuse.cpp [hunk 4]
```diff
@@ -3718,8 +3718,9 @@ int iuse::granade_act(player *, item *it, bool t, const tripoint &pos)
 
 int iuse::c4(player *p, item *it, bool, const tripoint& )
 {
-    int time = query_int(_("Set the timer to (0 to cancel)?"));
-    if (time <= 0) {
+    int time;
+    bool got_value = query_int( time, _("Set the timer to (0 to cancel)?") );
```
Comment (mugling) [inline src/iuse.cpp]: astyle

**Comment by mugling** [comment]
Unused variable

File: src/output.cpp [hunk 5]
```diff
@@ -700,17 +700,25 @@ bool internal_query_yn( const char *mes, va_list ap )
...
+    int num = atoi( raw_input.c_str() );
+    if( raw_input != "0" && num == 0) { //invalid input
```
Comment (mugling) [inline src/output.cpp]: Won't work for `00` which although questionable should be accepted as per Postels law. Consider `strtol`

**Pull request merged**